### ISSUE TITLE
RTMP Push fails if stream key contains ampersand (&)

### ISSUE THREAD

**Issue by bchah** [AirenSoft/OvenMediaEngine#704]
It looks like you cannot push an RTMP stream from OME if there is an ampersand <&> character in the stream key.

For example, a SignedPolicy URL (when streaming OME to OME) would fail like this and the Push API returns an 'error' state:

`[2022-03-19 08:22:56.564] E [SPAPIServer-T80:1938] RTMPWriter | rtmp_writer.cpp:140  | Error opening file. error(-5:Input/output error), filename(rtmp://myome/live/mystream?policy=x&signature=y)`

But a similar URL (including ? and = characters, but no &) works no problem. We did some testing on our own OME instances but also tried with your latest Docker image and the default Server.xml configuration (assuming <RTMPPush> is activated in the publisher).

Doing some research into FFMPEG, the *avio_open2* function seems to be the one that parses the URL. I don't have the C++ skills to fix this and we've tried encoding the <&> character every which way (%26, \&, &&, etc...) with no success.

Has anybody else seen this? It would be great to solve since OME can push to lots of places like YouTube, etc... just not to another OME! 😅

**Comment by Keukhan** [comment]
@bchah 

This is a great report. I will check the problem and reply to you.

Thanks.

**Comment by getroot** [comment]
@bchah 
Are there cases where the RTMP stream key contains & on YouTube or other streaming platforms?

**Comment by Keukhan** [comment]
@bchah 

I tried to reproduce it to check the problem. request a push by including the &(Ampersand) character in the target URL but, there was no problem.

OME Push to OME
> {
  "id" : "test_ome",
  "stream": {
    "name": "stream_o"
  },
  "protocol": "rtmp",
  "url":"rtmp://192.168.0.160:21935/app",
  "streamKey":"stream_2?policy=eyJ1cmxfZXhwaXJlIjoxNzA0Mzc3NTIwMTc4fQ&signature=CPhZX7TcoFgiY3s9YmymzmDmEOY"
}

with RTMP Push, OME -> OME works normally even if the & character is included in the StreamKey.

OME to YouTube

> {
>   "id" : "test_youtube",
>   "stream": {
>     "name": "stream_o"
>   },
>   "protocol": "rtmp",
>   "url":"rtmp://a.rtmp.youtube.com/live2",
>   "streamKey":"xxxx-xxxx-tepj-91qa-4yft&test=1"
> }

I also tried the following on youtube, but I can't connect. OBS is the same symptom. Apparently, if there are additional parameters, access is restricted.

There seems to be no problem in my opinion, but if there are other use cases that are problematic, please provide additional comments.

Thanks.  👍

**Comment by Keukhan** [comment]
@bchah

I tried to reproduce it to check the problem. request a push by including the &(Ampersand) character in the target URL but, there was no problem.

OME Push to OME


```
{
  "id" : "test_ome",
  "stream": {
    "name": "stream_o"
  },
  "protocol": "rtmp",
  "url":"rtmp://192.168.0.160:21935/app",
  "streamKey":"stream&2"
}
```

or

```
{
  "id" : "test_ome",
  "stream": {
    "name": "stream_o"
},
  "protocol": "rtmp",
  "url":"rtmp://192.168.0.160:21935/app",
  "streamKey":"stream_2?policy=eyJ1cmxfZXhwaXJlIjoxNzA0Mzc3NTIwMTc4fQ&signature=CPhZX7TcoFgiY3s9YmymzmDmEOY"
}
```


with RTMP Push, OME -> OME works normally even if the & character is included in the StreamKey.

OME to YouTube

```
 {
   "id" : "test_youtube",
   "stream": {
   "name": "stream_o"
 },
   "protocol": "rtmp",
   "url":"rtmp://a.rtmp.youtube.com/live2",
   "streamKey":"xxxx-xxxx-tepj-91qa-4yft?test=1"
 }
```

I also tried the following on youtube, but I can't connect. OBS is the same symptom. Apparently, if there are additional parameters, access is restricted.

There seems to be no problem in my opinion, but if there are other use cases that are problematic, please provide additional comments.

Thanks. 👍

**Comment by bchah** [comment]
@Keukhan Thank you for testing - it's odd that you can't reproduce it on your side, but I can't overcome it on my side.

I tried setting up two brand new OME installs and pushing from one to another. I also noticed you were testing on LAN so I did the same. On the push side, the API returns the same state of "error" and this appears in the log:

```
[2022-03-22 13:39:55.305] E [AW-RTMPPush1:546] RTMPWriter | rtmp_writer.cpp:140  | Error opening file. error(-5:Input/output error), filename(rtmp://192.168.0.133:1935/live/helloWorld?p=eyJ1cmxfZXhwaXJlIjo3MjQ1NTk0MTMwMDAwfQ&s=V7iy2b-a76NPgLbBuMpVgwkd__Q)
[2022-03-22 13:40:00.294] W [AW-RTMPPush1:546] RTMPPush | rtmppush_session.cpp:86   | Could not supported codec. track_id:2, codec_id: 8
```

Of course this might look like I am sending the wrong codecs but the same stream works fine publishing to OME without a policy + signature argument attached, or YouTube, etc... So I think the codec error is a red herring, especially because the request does actually arrive to the remote server but it is interpreting the signature in a strange way:

```
[2022-03-22 13:39:55.360] I [SPRTMP-T1935:529] RTMPProvider | rtmp_provider.cpp:152  | A RTMP client has connected from <ClientSocket: 0x7fddbc0016c0, #19, Connected, TCP, Nonblocking, 192.168.0.188:54326>
[2022-03-22 13:39:55.366] W [SPRTMP-T1935:529] RTMPProvider | rtmp_stream.cpp:340  | Signature value is invalid(expected : V7iy2b-a76NPgLbBuMpVgwkd__Q | input : V7iy2b-a76NPgLbBuMpVgwkd__Q/helloWorld).
[2022-03-22 13:39:55.367] I [SPRTMP-T1935:529] Provider | stream.cpp:60   | Unknown/(19) has been stopped playing stream
[2022-03-22 13:39:55.367] E [SPRTMP-T1935:529] RTMPProvider | rtmp_stream.cpp:315  | SendAmfCreateStreamResult Fail
[2022-03-22 13:39:55.367] I [SPRTMP-T1935:529] RTMPProvider | rtmp_provider.cpp:181  | The RTMP client has disconnected: [Unknown/], remote: <ClientSocket: 0x7fddbc0016c0, #19, Closed, TCP, Nonblocking, 192.168.0.188:54326>
```

As you can see the signature is correct but on the receiving end it thinks the stream key has been placed after the signature? Could this actually be an issue with SignedPolicy and the way it receives the arguments when the stream is being pushed from OME? I hope this provides a further clue.

**Comment by bchah** [comment]
Thank you @Keukhan! I could not test this on the Master branch because the API does not seem to work with the new HTTP 2.0 additions. But I pulled the fix back into the 0.13.2 release and... it works!! 🎉

![Screen Shot 2022-03-22 at 5 56 37 PM](https://user-images.githubusercontent.com/33604674/159583090-9b0823a3-15cf-4c68-bc01-32f90b5e47a7.jpg)

Thank you for your quick effort to solve this!

**Comment by Keukhan** [comment]
👍
